File: leetcode/0130-surrounded-regions/0130-surrounded-regions.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solve = function (board) {
    // 겉에를 돌면서 O인 부분을 찾고, DFS를 시행
    // 그렇게 DFS를 시행한 이후 체크되지 않은 O는 전부 X로 변경

    const N = board.length;
    const M = board[0].length;

    const visited = Array.from({ length: M }, () => Array(N).fill(false));

    // 겉에 돌기
    for (let n = 0; n < N; n++) {
        DFS(n, 0)

        if (n === 0 || n === N - 1) {
            // 맨 위 혹은 맨 아래 탐색
            for (let m = 1; m < M - 1; m++) {
                // dfs
                DFS(n, m)
            }
        }

        DFS(n, M - 1)
    }

    // console.log(visited)

    for(let n = 0; n < N; n++) {
        for(let m = 0; m < M; m++) {
            if(!visited[n][m]) {
                board[n][m] = 'X'
            }
        }
    }

    function DFS(sr, sc) {
        if(board[sr][sc] === 'X') {
            return;
        }

        if (visited[sr][sc]) {
            return;
        }

        const stack = [[ sr, sc ]];
        const DIR = [[0, 1], [0, -1], [1, 0], [-1, 0]];

        while (stack.length) {
            const [row, col] = stack.pop();
            visited[row][col] = true;

            for (let i = 0; i < DIR.length; i++) {
                let newRow = row + DIR[i][0];
                let newCol = col + DIR[i][1];

                // check edge
                if(
                    N <= newRow || newRow < 0 ||
                    M <= newCol || newCol < 0
                ) {
                    continue;
                }
 
                // check visited
                if (visited[newRow][newCol]) {
                    continue;
                }

                // check X
                if (board[newRow][newCol] === 'X') {
                    continue;
                }

                visited[newRow][newCol] = true;
                stack.push([newRow, newCol])
            }
        }
    }
};

File: leetcode/0130-surrounded-regions/0130-surrounded-regions.swift
class Solution {
    func solve(_ board: inout [[Character]]) {
        let N = board.count
        let M = board[0].count

        let DIR_LIST = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        var visited = Array(
            repeating: Array(repeating: false, count: M)
            , count: N
        )
        var stack: Array<Array<Int>> = Array();
        for n in 0...N - 1 {
            // 맨 왼쪽, 오른쪽
            stack.append([n, 0])
            stack.append([n, M - 1])

            // 중간 
            if((n == 0 || n == N - 1) && M > 2) {
                for m in 1...M - 2 {
                    stack.append([n, m])
                }
            }
        }

        // print("finished setting stack")
        // print("stack: \(stack)");

        // DFS
        while(stack.count > 0) {
            guard let pos = stack.popLast() else { continue }
            let n = pos[0]
            let m = pos[1]

            if(board[n][m] == "X") {
                continue
            }

            visited[n][m] = true

            // print("Max N, M: \(N - 1), \(M - 1)")
            for dir in DIR_LIST {
                let newN = n + dir[0]
                let newM = m + dir[1]

                // print(newN, newM)

                // check "out"
                if(
                    N <= newN || newN < 0
                ||  M <= newM || newM < 0
                ) {
                    // print("out")
                    continue
                }

                // check "visited"
                if(visited[newN][newM]) { 
                    // print("visited")
                    continue 
                }
                
                // check "X"
                if(board[newN][newM] == "X") {
                    // print("X")
                    continue 
                }

                visited[newN][newM] = true;
                stack.append([newN, newM])
            }
        }

        for (n, col) in board.enumerated() {
            for (m, row) in col.enumerated() {
                if(!visited[n][m]) {
                    board[n][m] = "X"
                }
            }
        }
    }
}